src/test/java/org/elasticsearch/indices/IndicisServiceTest.java

```diff
@@ -0,0 +1,130 @@
+        assertTrue("shard is removed", indicesService.canDeleteShardContent(new ShardId("test", 0), meta));
+    }
+
+    public void testDeleteIndexStore() throws Exception {
+        IndicesService indicesService = getIndicesService();
+        IndexService test = createIndex("test");
+        ClusterService clusterService = getInstanceFromNode(ClusterService.class);
+        IndexMetaData firstMetaData = clusterService.state().metaData().index("test");
+        assertTrue(test.hasShard(0));
+
+        try {
+            indicesService.deleteIndexStore("boom", clusterService.state(), firstMetaData);
+            fail();
+        } catch (ElasticsearchIllegalStateException ex) {
+            // all good
+        }
+
+        GatewayMetaState gwMetaState = getInstanceFromNode(GatewayMetaState.class);
+        MetaData meta = gwMetaState.loadMetaState();
+        assertNotNull(meta);
+        assertNotNull(meta.index("test"));
+        assertAcked(client().admin().indices().prepareDelete("test"));
+
+        meta = gwMetaState.loadMetaState();
+        assertNotNull(meta);
+        assertNull(meta.index("test"));
+
+
+        createIndex("test");
+        client().prepareIndex("test", "type", "1").setSource("field", "value").setRefresh(true).get();
+        client().admin().indices().prepareFlush("test").get();
+        assertHitCount(client().prepareSearch("test").get(), 1);
+        IndexMetaData secondMetaData = clusterService.state().metaData().index("test");
+        assertAcked(client().admin().indices().prepareClose("test"));
+        NodeEnvironment nodeEnv = getInstanceFromNode(NodeEnvironment.class);
+        Path[] paths = nodeEnv.shardDataPaths(new ShardId("test", 0), clusterService.state().getMetaData().index("test").getSettings());
+        for (Path path : paths) {
+            assertTrue(Files.exists(path));
+        }
+
+        try {
+            indicesService.deleteIndexStore("boom", clusterService.state(), secondMetaData);
+            fail();
+        } catch (ElasticsearchIllegalStateException ex) {
+            // all good
+        }
+
+        for (Path path : paths) {
+            assertTrue(Files.exists(path));
+        }
+
+        // now delete the old one and make sure we resolve against the name
+        try {
+            indicesService.deleteIndexStore("boom", clusterService.state(), firstMetaData);
+            fail();
+        } catch (ElasticsearchIllegalStateException ex) {
+            // all good
+        }
+        ClusterState fakeState = ClusterState.builder(clusterService.state().getClusterName()).build();
+        indicesService.deleteIndexStore("boom", fakeState, firstMetaData);
+
+        assertAcked(client().admin().indices().prepareOpen("test"));
+        ClusterHealthResponse resp = client().admin().cluster()
+                .health(Requests.clusterHealthRequest("test").waitForEvents(Priority.LANGUID)).actionGet();
+        assertEquals(resp.getStatus(), ClusterHealthStatus.RED);
```

Comment:
maybe I miss it and it's somewhere else - do we want to make sure we can't delete the content if there is some reference to the shard folder (but no new index)?